### PR TITLE
Adds a number of flags/options

### PR DESCRIPTION
**Pull request by valdisrigdon**
These flags are required to comply with various parts of the CIS Kubernetes Benchmark.

The ability to disable basic and token auth are parts of the CIS Kubernetes Benchmark, but that isn't as simple as a flag, so it's added as a second commit.

### CODE REVIEW

**Comment by valdisrigdon** [comment]
CLA is signed.

File: pkg/apis/kops/componentconfig.go [hunk 1]
```diff
@@ -156,6 +156,14 @@ type KubeletConfigSpec struct {
...
 	// Tells the Kubelet to fail to start if swap is enabled on the node.
 	FailSwapOn *bool `json:"failSwapOn,omitempty" flag:"fail-swap-on"`
+	// EventQps If > 0, limit event creations per second to this value. If 0, unlimited.
+	EventQPS *int32 `json:"eventQps,omitempty" flag:"event-qps"`
+	// MakeIptablesUtilChains will ensure iptables utility rules are present on host.
+	MakeIptablesUtilChains *bool `json:"makeIptablesUtilChains,omitempty" flag:"make-iptables-util-chains"`
+	// CAdvisorPort The port of the localhost cAdvisor endpoint (set to 0 to disable) (default 4194)
+	CAdvisorPort *int32 `json:"cadvisorPort,omitempty" flag:"cadvisor-port"`
+	// ProtectKernelDefaults Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.
+	ProtectKernelDefaults *bool `json:"protectKernelDefaults,omitempty" flag:"protect-kernel-defaults"`
```
Comment (mikesplain) [inline pkg/apis/kops/componentconfig.go]: Should any of these be documented? None jump out to me as something most users would adjust...

File: pkg/apis/kops/componentconfig.go [hunk 2]
```diff
@@ -196,6 +196,10 @@ type KubeProxyConfig struct {
 type KubeAPIServerConfig struct {
 	// Image is the docker container used
 	Image string `json:"image,omitempty"`
+	// DisableBasicAuth removes the --basic-auth-file flag
+	DisableBasicAuth bool `json:"disableBasicAuth,omitempty"`
+	// EnableTokenAuth removes the --token-auth-file flag
+	DisableTokenAuth bool `json:"disableTokenAuth,omitempty"`
```
Comment (mikesplain) [inline pkg/apis/kops/componentconfig.go]: Can we get docs for these included in this commit?

**Comment by mikesplain** [comment]
Looks pretty good, thanks for the contribution @valdisrigdon!

/assign

File: pkg/apis/kops/componentconfig.go (hunk 2 above)
Comment (valdisrigdon) [reply pkg/apis/kops/componentconfig.go]: I've updated the commit to include docs for these two options.

**Comment by mikesplain** [comment]
Great thanks so much @valdisrigdon! 
/lgtm

**Comment by chrislovecnm** [comment]
We are close to 1.9, so we are probably going to hold this for 1.10.  The 1.10 is going to be a pretty quick turn around.

**Comment by mikesplain** [comment]
Agreed @chrislovecnm :)

**Comment by mikesplain** [comment]
/assign @chrislovecnm 

Mind taking another look at this?

**Comment by chrislovecnm** [comment]
Can we get this rebased? Ping me on slack and I will take a look after a rebase.  Thanks

**Comment by kaazoo** [comment]
Any update?

**Comment by valdisrigdon** [comment]
@chrislovecnm This is rebased.

**Comment by mikesplain** [comment]
readding /lgtm

**Comment by valdisrigdon** [comment]
Bump.

**Comment by valdisrigdon** [comment]
@chrislovecnm Is there an interest in getting this merged? I've rebased a few times and about 6 weeks since 1.9 was cut. I'm happy to rebase again if you'd like to get this in.

File: pkg/apis/kops/componentconfig.go [hunk 3]
```diff
@@ -164,6 +164,14 @@ type KubeletConfigSpec struct {
 	FailSwapOn *bool `json:"failSwapOn,omitempty" flag:"fail-swap-on"`
 	// ExperimentalAllowedUnsafeSysctls are passed to the kubelet config to whitelist allowable sysctls
 	ExperimentalAllowedUnsafeSysctls []string `json:"experimentalAllowedUnsafeSysctls,omitempty" flag:"experimental-allowed-unsafe-sysctls"`
+	// EventQps If > 0, limit event creations per second to this value. If 0, unlimited.
+	EventQPS *int32 `json:"eventQps,omitempty" flag:"event-qps"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: So we now (finally) have more componentconfig upstream in kubernetes/kubernetes.  We should try to be consistent where it's possible to smooth the path for users.

Looking at componentconfig, the json form should probably be capitalized as `QPS`, and in fact should be `eventRecordQPS`

(`pkg/kubelet/apis/kubeletconfig/v1beta1/types.go` has `EventRecordQPS *int32 `json:"eventRecordQPS,omitempty"``

File: pkg/apis/kops/componentconfig.go [hunk 4]
```diff
@@ -315,9 +327,14 @@ type KubeAPIServerConfig struct {
 	FeatureGates map[string]string `json:"featureGates,omitempty" flag:"feature-gates"`
 	// MaxRequestsInflight The maximum number of non-mutating requests in flight at a given time.
 	MaxRequestsInflight int32 `json:"maxRequestsInflight,omitempty" flag:"max-requests-inflight" flag-empty:"0"`
-
 	// EtcdQuorumRead configures the etcd-quorum-read flag, which forces consistent reads from etcd
 	EtcdQuorumRead *bool `json:"etcdQuorumRead,omitempty" flag:"etcd-quorum-read"`
+	// Enable profiling via web interface
+	Profiling *bool `json:"profiling,omitempty" flag:"profiling"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: For consistency, `enableProfiling`

File: pkg/apis/kops/componentconfig.go [hunk 5]
```diff
@@ -315,9 +327,14 @@ type KubeAPIServerConfig struct {
...
+	// Verify service account token
+	ServiceAccountLookup *bool `json:"serviceAccountLookup,omitempty" flag:"service-account-lookup"`
+	// Repair malformed requests from clients
+	RepairMalformedUpdates *bool `json:"repairMalformedUpdates,omitempty" flag:"repair-malformed-updates"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: Do we need this one?  It's deprecated and defaults to false now.

File: pkg/apis/kops/componentconfig.go [hunk 6]
```diff
@@ -379,6 +396,8 @@ type KubeControllerManagerConfig struct {
 	HorizontalPodAutoscalerUseRestClients *bool `json:"horizontalPodAutoscalerUseRestClients,omitempty" flag:"horizontal-pod-autoscaler-use-rest-clients"`
 	// FeatureGates is set of key=value pairs that describe feature gates for alpha/experimental features.
 	FeatureGates map[string]string `json:"featureGates,omitempty" flag:"feature-gates"`
+	// Enable profiling via web interface
+	Profiling *bool `json:"profiling,omitempty" flag:"profiling"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: `enableProfiling` here also I think

File: pkg/apis/kops/componentconfig.go [hunk 7]
```diff
@@ -415,6 +434,8 @@ type KubeSchedulerConfig struct {
 	LogLevel int32 `json:"logLevel,omitempty" flag:"v"`
 	// Image is the docker image to use
 	Image string `json:"image,omitempty"`
+	// Enable profiling via web interface
+	Profiling *bool `json:"profiling,omitempty" flag:"profiling"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: `enableProfiling` :-)

File: pkg/apis/kops/v1alpha1/componentconfig.go [hunk 8]
```diff
@@ -164,6 +164,14 @@ type KubeletConfigSpec struct {
 	FailSwapOn *bool `json:"failSwapOn,omitempty" flag:"fail-swap-on"`
 	// ExperimentalAllowedUnsafeSysctls are passed to the kubelet config to whitelist allowable sysctls
 	ExperimentalAllowedUnsafeSysctls []string `json:"experimental_allowed_unsafe_sysctls,omitempty" flag:"experimental-allowed-unsafe-sysctls"`
+	// EventQps If > 0, limit event creations per second to this value. If 0, unlimited.
+	EventQPS *int32 `json:"eventQps,omitempty" flag:"event-qps"`
```
Comment (justinsb) [inline pkg/apis/kops/v1alpha1/componentconfig.go]: Same naming changes for the json fields needed in v1alpha1 and v1alpha2, I think!  I won't repeat them so as not to flood you with comments!

File: pkg/apis/kops/componentconfig.go [hunk 9]
```diff
@@ -164,6 +164,14 @@ type KubeletConfigSpec struct {
 	FailSwapOn *bool `json:"failSwapOn,omitempty" flag:"fail-swap-on"`
 	// ExperimentalAllowedUnsafeSysctls are passed to the kubelet config to whitelist allowable sysctls
 	ExperimentalAllowedUnsafeSysctls []string `json:"experimentalAllowedUnsafeSysctls,omitempty" flag:"experimental-allowed-unsafe-sysctls"`
+	// EventQps If > 0, limit event creations per second to this value. If 0, unlimited.
+	EventQPS *int32 `json:"eventQps,omitempty" flag:"event-qps"`
+	// MakeIptablesUtilChains will ensure iptables utility rules are present on host.
+	MakeIptablesUtilChains *bool `json:"makeIptablesUtilChains,omitempty" flag:"make-iptables-util-chains"`
```
Comment (justinsb) [inline pkg/apis/kops/componentconfig.go]: For consistency, can we make this `makeIPTablesUtilChains` also?

**Comment by justinsb** [comment]
Sorry this took so long.  So yes, these are great changes, though I think because we now have componentconfig in kubernetes/kubernetes we should make the JSON field names the same as in componentconfig.  (The go field names don't matter - they aren't user-facing!)

In general if you break this up into separate PRs we can get them merged much more quickly, because they would only be one idea.  But I'm also trying to get better on review turnaround time...

LGTM with the json field names per upstream.

**Comment by bzuelke** [comment]
Agreed would like to see if there was a road map for this merge. This would help out a ton for getting the clusters to CIS standards

**Comment by nimble-software** [comment]
Any progress on this yet?

**Comment by justinsb** [comment]
So I'm going to move to 1.11, and hopefully we can get this (or similar functionality) merged in 1.11.  But it's too late for 1.10

**Comment by nimble-software** [comment]
Ok thanks for the update. Is there a proposed date when 1.11 will be released or is it still to early to tell?

**Comment by noqcks** [comment]
Looks like the json field names still need to be updated by @valdisrigdon before we can merge?

**Comment by fernandocarletti** [comment]
@justinsb I think @valdisrigdon is not maintaining this PR anymore. @valdisrigdon would be fine if I take the current code and open a new Pull Request, I see that the Appian fork does not have the branch anymore (it was the source of this PR, right?). I would be glad to keep working on this since the flags are needed by many (including me). Would it be ok?

**Comment by johannes-gehrs** [comment]
This PR would be important for my org. Let me know if I can help getting this merged (as a non-KOPS-committer).

**Comment by jaybe78** [comment]
how long before merging this ? can someone give an update ?

**Comment by mat285** [comment]
Any updates on this?

**Comment by mpfgomes** [comment]
+1

**Comment by msiuts** [comment]
Are there any updates on this? Something one could do to speed this up?

**Comment by javierlgadou** [comment]
Any updates? I'm doing some hardening to my cluster

**Comment by olemarkus** [comment]
With #8967 merged and no activity here for quite some time, I guess this one can be closed now.

**Comment by mikesplain** [comment]
Thanks @olemarkus agreed.

/close

**Pull request closed without merging**